File: Spark/codes/SparkSQL/java/dataframe/DataFrameRDDProgram.java
package dataframe;

import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

import java.util.ArrayList;
import java.util.List;

public class DataFrameRDDProgram {



    public static void main(String[] args) {
        String inputPath = "file:///home/willhope/app/data/infos.txt";

        //2. 设置SparkSession，master
        SparkSession sparkSession = SparkSession.builder().master("local[2]").appName("DataFrameRDDProgram").getOrCreate();

        //3. 读取文件，生成RDD
        JavaRDD<String> javaRDD = sparkSession.sparkContext().textFile(inputPath,1).toJavaRDD();

        //4. 处理原始数据RDD，生成新的RDD
        JavaRDD<Row> rowJavaRDD = javaRDD.map(
                line -> {
                    String[] infos = line.split(",");
                    Integer id = Integer.parseInt(infos[0]);
                    String name = infos[1];
                    Integer age = Integer.parseInt(infos[2]);
                    Row row = RowFactory.create(id, name, age);
                    return row;
                }
        );

        //5. 创建自定义数据类型
        List<StructField> fields = new ArrayList<>();
        StructField field_id = DataTypes.createStructField("id",DataTypes.IntegerType,true);
        StructField field_name = DataTypes.createStructField("name", DataTypes.StringType, true);
        StructField field_age = DataTypes.createStructField("age", DataTypes.IntegerType, true);
        fields.add(field_id);
        fields.add(field_name);
        fields.add(field_age);
        //6. 将自定义的数据类型StructType化，生成schema
        StructType schema = DataTypes.createStructType(fields);

        //7. 把RDD和schema关联起来并查询
        Dataset<Row> personDataFrame = sparkSession.createDataFrame(rowJavaRDD, schema);
        personDataFrame.printSchema();
        personDataFrame.show();
        personDataFrame.filter(personDataFrame.col("age").$greater(20)).show();

        //创建临时表，进行sql操作
//        personDF.createOrReplaceGlobalTempView("t_person");
//        sparkSession.sql("select * from t_person order by age desc").show();

        //关闭资源
        sparkSession.stop();

    }


}

File: Spark/codes/SparkSQL/scala/dataframe/DataFrameRDDProgram.scala
package org.example.dataframe

import org.apache.spark.sql.{Row, SparkSession}
import org.apache.spark.sql.types.{IntegerType, StringType, StructField, StructType}

object DataFrameRDDProgram {

  def main(args: Array[String]): Unit = {

    //1. 设置文件路径
    val inputPath = "file:///home/willhope/app/data/infos.txt"

    //2. 设置SparkSession，master
    val spark = SparkSession.builder().appName("DataFrameRDDProgram").master("local[2]").getOrCreate()

    //3. sparkContext可以获取到rdd，textfile可以从hdfs上读取文件，或者从本地获取文件
    val rdd = spark.sparkContext.textFile(inputPath)

    //4. 处理文件内容，不需要像反射方法中定义类，使用Row接收数据即可
    val infoRDD = rdd.map(_.split(",")).map(line => Row(line(0).toInt, line(1), line(2).toInt))

    //5. 设置structType，设置Row数据的结构，第三个参数表示能否为空
    val structType = StructType(Array(StructField("id",IntegerType,true),
      StructField("name",StringType,true),
      StructField("age",IntegerType,true)
    ))

    //6. 将rdd转换为Dataframe
    val infoDF = spark.createDataFrame(infoRDD,structType)
    infoDF.printSchema()
    infoDF.show()

    infoDF.filter(infoDF.col("age") > 30).show()

    //如果对编程不了解，也可以采用sql的方式来解决，创建一张临时的表，然后可以直接使用spark-sql
//    infoDF.createOrReplaceTempView("infos")
//    spark.sql("select * from infos where age>30").show()
    spark.stop()
  }

}
